lib/translator/mode_sense.cc

```diff
@@ -0,0 +1,390 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mode_sense.h"
+
+namespace translator {
+
+namespace {
+
+// Section 6.3.3.1
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+constexpr scsi::CachingModePage caching_mode_page = {
+    .ps = 0,
+    .spf = 0,
+    .page_code = scsi::ModePageCode::kCacheMode,
+    .page_length = 0x12,
+    .ic = 1,
+};
+
+// Section 6.3.3.2
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+constexpr scsi::ControlModePage control_mode_page = {
+    .ps = 0,
+    .spf = 0,
+    .page_code = scsi::ModePageCode::kControlMode,
+    .page_length = 0x0a,
+    .tst = 0,
+    .tmf_only = 0,
+    .dpicz = 0,
+    .d_sense = 1,
+    .gltsd = 1,
+    .rlec = 0,
+    .qam = 0,
+    .nuar = 0,
+    .qerr = 0x01,
+    .rac = 0,
+    .ua_intlck_ctrl = 0,
+    .swp = 0,
+    .ato = 0,
+    .tas = 0x01,
+    .autoload_mode = 0,
+    .busy_timeout_period = 0xFFFF,
+    .estct = 0};
+
+// Section 6.3.3.3
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+constexpr scsi::PowerConditionModePage power_condition_mode_page = {
+    .ps = 0,
+    .spf = 0,
+    .page_code = scsi::ModePageCode::kPowerConditionMode,
+    .page_length = 0x26,
+    .pm_bg_precedence = 0};
+
+struct CommonCmdAttributes {
+  scsi::ModePageCode page_code;
+  scsi::PageControl pc;
+  bool dbd;
+  bool llbaa;
+};
+
+// Generates an nvme identify command for getting block descriptor data
+StatusCode GenerateBlockDescriptorIdentifyCmd(
+    nvme::GenericQueueEntryCmd& nvme_cmd, Allocation& allocation,
+    uint32_t nsid) {
+  nvme_cmd = {.opc = static_cast<uint8_t>(nvme::AdminOpcode::kIdentify),
+              .nsid = nsid};
+  uint64_t prp_ptr = AllocPages(1);
+  if (prp_ptr == 0) return StatusCode::kFailure;
+  allocation.data_addr = prp_ptr;
+  allocation.data_page_count = 1;
+  nvme_cmd.dptr.prp.prp1 = prp_ptr;
```

Comment:
https://github.com/googleinterns/scsi2nvme/pull/116 has been merged so the SetPages can be used